Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift

```diff
@@ -139,16 +139,37 @@ public final class DefaultSettingsProvider: DefaultSettingsProviding {
             essentialKeys: DefaultSettingsProvider.essentialTargetSettings,
             newXcodeKeys: DefaultSettingsProvider.xcodeVersionSpecificSettings
         )
+        let mergeableSettings = mergeableSettings(for: target, configuration: buildConfiguration)
         var settings: SettingsDictionary = [:]
         settingsHelper.extend(buildSettings: &settings, with: targetDefaultAll)
         settingsHelper.extend(buildSettings: &settings, with: additionalTargetDefaults)
         settingsHelper.extend(buildSettings: &settings, with: targetDefaultVariant)
+        settingsHelper.extend(buildSettings: &settings, with: mergeableSettings)
         settingsHelper.extend(buildSettings: &settings, with: projectOverridableTargetDefaultSettings(for: project))
         return settings.filter(filter)
     }
 
     // MARK: - Private
 
+    private func mergeableSettings(for target: Target, configuration: BuildConfiguration) -> SettingsDictionary {
+        var settings: SettingsDictionary = [:]
+        if target.mergeable {
+            if target.product != .dynamicLibrary {
+                fatalError("Only products that are dynamic libraries can be marked as mergeable")
+            }
```

Comment:
As a suggestion, perhaps to help avoid this, to declare target as mergeable we could introduce a new product type `.mergableLibrary`  which under the hood would map to the appropriate Xcode product type (be it `.dynamicLibary ` or `.framework`?)

Note: this `fatalError` won't produce a user friendly error when generating projects - it will manifest as a crash  :/

Reply:
> As a suggestion, perhaps to help avoid this, to declare target as mergeable we could introduce a new product type .mergableLibrary which under the hood would map to the appropriate Xcode product type (be it .dynamicLibary  or .framework?) - alternatively we'd need to add a lint rule to fail in case a target is mergeable but isn't one of the supported product types.

I see the value in doing this and letting the compiler do the validation. I wonder if those two products are the only ones that will support mergeable and therefore the list of additional `.mergeableXX` product types that we have to add are minimal. If so, I'm inclined to follow @kwridan's suggestion.

Regarding the error handling, the convention that we follow is raising [`FatalError`s](https://github.com/tuist/tuist/blob/main/Sources/TuistSupport/Errors/FatalError.swift#L45) because we have an error handler at the root that takes care of formatting the errors appropriately.

Reply:
I agree, as we have `dynamicLibrary` and `staticLibrary`, I think it would make sense to have `mergeableLibrary`.
Wouldn't do it now, but in the future it could make even more sense to have `library(.dynamic/static/mergeable)`, so we don't explode the number of cases in the outer enum